Blogpost for the PyTorch blog

lezcano opened the pull request.

rgommers commented on blogpost/post.md:
```diff
@@ -0,0 +1,252 @@
+this direction. Quansight and Meta continue working in this direction,
+improving the compatibility between PyTorch and the rest of the ecosystem.
+
+From Quansight, we would like to thank Meta for funding this project and all
```
I suggest to change "and all the previous work ..." to "as well as previous work on improving NumPy compatibility within PyTorch".

rgommers commented on blogpost/post.md:
```diff
@@ -0,0 +1,252 @@
+
+From Quansight, we would like to thank Meta for funding this project and all
+the previous work that lead to it, like improving the NumPy compatibility
+within PyTorch, and developing the [python Array API](https://data-apis.org/array-api/latest/).
```
This didn't involve any Meta funding actually. Quansight and Intel have been the largest funders, the others are listed on https://data-apis.org/. Meta did fund the work on adding PyTorch support to https://github.com/data-apis/array-api-compat, but that work and funding acknowledgement are covered in Thomas' blog post on the scikit-learn work.

rgommers reviewed (commented):
Thanks Mario, great post.

lezcano commented on blogpost/post.md:
```diff
@@ -0,0 +1,252 @@
+NumPy code in CUDA, we can simply modify this code to call `x.numpy()` and when
+running it under `device("cuda")`, as we did above, it will generate efficient
+CUDA code from original NumPy calls without copying the data from CUDA to CPU
+at all.
```
I was debating about adding it or not. I will.

lezcano commented on blogpost/post.md:
```diff
@@ -0,0 +1,252 @@
+
+## Beyond NumPy: SciPy and scikit-learn
+
+In parallel to this effort, other Quansight engineers have designed, proposed
```
Re other. I agree. I want to start mentioning Quansight here, but it seems a bit out of the blue. Perhaps also mention it at in the first paragraph of the blogpost?

lezcano commented:
Addressed the review. The main missing point would be that I start talking about QS in the penultimate section. I would like to mention QS somewhere at the top as well, but I really like the introductory paragraph, and I don't know how. I've added a proposal in the last commit, but I don't fully like it. Could I get some feedback particularly on that?

lezcano commented:
In particular, could you guys give further feedback on the introductory paragraph and the last 2 sections? I'd like to make sure those are proper tight.

amjames commented on blogpost/post.md:
```diff
@@ -0,0 +1,263 @@
+The compiled function yields a 9x speed-up when running it on 1 core. Even
+better, since the compiled code also runs on multiple cores, we get a **57x speed-up**
+when running it on 32 cores. Note that vanilla NumPy always runs on
+just one core.
```
Have you compared the performance to other options for compiling numpy. The `get_labels` example contains functions which are supported by `numba`, that might be an interesting datapoint to see how the `torch.compile` speedup compares to `numba.jit`

lezcano replied:
Not really, as this would just make our announcement post way too long. I will expect other people to put up posts comparing this and other approaches (julia / numba / torch.jit / mojo), but that's beyond the scope of this post IMO

ev-br replied:
I plan to write a follow-up, based on https://github.com/Quansight-Labs/numpy_pytorch_interop/pull/168 or https://github.com/Quansight-Labs/numpy_pytorch_interop/tree/main/e2e/smoke, comparing to numba.jit may fit there.

lezcano replied:
That'd be super cool!

ev-br replied:
Weirdly enough, throwing `@numba.njit` in place of `torch.compile` in this example runs into several rough edges of numba:

- `np.linalg.norm(..., axis=2)` does not compile, chokes on `axis`.
-  Replacing `np.linalg.norm` with 

```
def norm(a, axis):
    s = (a.conj() * a).real
    return np.sqrt(s.sum(axis=axis))
```

and `njit`-ting both `norm`  and `get_labels` gives a _slowdown_ of about 60%.

- trying `@njit(parallel=True)` on `get_labels` crashes the compiler, again on the `axis` argument (`TypeError: Failed in nopython mode pipeline (step: Preprocessing for parfors) got an unexpected keyword argument 'axis'`)

- trying `@njit(parallel=True)` on the `norm` only and `@njit` on `get_labels` yields a slowdown w.r.t. numpy of 'only' 20%. 


Got to admit I never had much luck with numba in non-toy situations.

ev-br commented on blogpost/post.md:
```diff
@@ -0,0 +1,261 @@
+# Compiling NumPy code into C++ or CUDA via `torch.compile`
+
+Quansight engineers have implemented support for tracing through NumPy code via
```
Stictly speaking, QS and Meta engineers (Mengwei, Voz). <hairsplitter mode off>

ev-br commented on blogpost/post.md:
```diff
@@ -0,0 +1,261 @@
+Quansight engineers have implemented support for tracing through NumPy code via
+`torch.compile` in PyTorch 2.1.
+This feature leverages PyTorch's compiler to generate efficient fused
+vectorized code without having to modify your original code. Even more, it
```
original NumPy code

ev-br commented on blogpost/post.md:
```diff
@@ -0,0 +1,261 @@
+## Compiling NumPy code into Parallel C++
+
+We take as our running example one step in a K-Means algorithm.
+This piece of code is borrowed from this [NumPy book](https://realpython.com/numpy-array-programming/#clustering-algorithms)
```
From the hair splitter gallery: the NumPy book is what Travis wrote back in the day, later to become the basis of the current numpy docs (I think)

ev-br commented on blogpost/post.md:
```diff
@@ -0,0 +1,261 @@
+We may inspect the generated C++ code by running the script with the
+environment variable `TORCH_LOGS=output_code`. When doing so, we can see that
+`torch.compile` was able to compile the broadcasting, together with the two
+reductions into just one for-loop, and parallelizes it using OpenMP
```
Not: was able to compile... and parallelize

ev-br commented on blogpost/post.md:
````diff
@@ -0,0 +1,261 @@
+        return tensor_fn(X, means)
+```
+
+This function now takes tensors in CUDA memory and returns tensors in CUDA
````
Maybe stress here that torch.compile deduces that the operations are all on CUDA and optimizes away to to/from_numpy() calls.

ev-br commented on blogpost/post.md:
```diff
@@ -0,0 +1,261 @@
+
+**Mixing NumPy and PyTorch**. In this example, we had to write a small adaptor
+to convert tensors to ndarrays and then back to tensors. In programs that mix PyTorch and
+NumPy this is already done by calling `x.detach().cpu().numpy()` (or simply
```
"This is already done" was not clear to me, done where and by whom (manually, automatically by dynamo)

ev-br commented on blogpost/post.md:
```diff
@@ -0,0 +1,261 @@
+running it under `device("cuda")`, as we did above, it will generate efficient
+CUDA code from original NumPy calls without copying the data from CUDA to CPU
+at all. Note that the resulting code would not run without `torch.compile`. For
+it to run in eager mode one would need to rollback to `x.numpy(force=True)`.
```
This paragraph I guess explains the previous one, but still it's confusing, at least to not very torch-savvy reader in me. Consider rephrasing?

ev-br commented on blogpost/post.md:
```diff
@@ -0,0 +1,261 @@
+PyTorch, is often used very differently. It is rather common to perform
+computations in NumPy and then do an if/else depending on values within the
+array, or perform operations in-place, perhaps via boolean masks. These
+constructions, while supported by `torch.compile`, hamper its performance.
```
Hairsplitting : these hamper the performance of compiled programs, not torch.compile itself

ev-br commented on blogpost/post.md:
```diff
@@ -0,0 +1,261 @@
+computations in NumPy and then do an if/else depending on values within the
+array, or perform operations in-place, perhaps via boolean masks. These
+constructions, while supported by `torch.compile`, hamper its performance.
+Changes like moving from in-place indexing to using `np.where`, writing the
```
Suggest ... from advanced indexing to using np.where --- ints and slices are OK, are they

lezcano replied:
Good catch, but not quite, it should be "modifying arrays in-place via indexing" or so. Meaning `x[idx] = t`. It doesn't matter if it's advanced indexing or not. I'll figure out how to improve the writing tho.

ev-br commented on blogpost/post.md:
```diff
@@ -0,0 +1,261 @@
+**Debugging NumPy code**. Debugging is rather tricky when a compiler is
+involved. To figure out whether an error you are hitting is a `torch.compile`
+error, or an error from the program, you can execute your NumPy program without
+`torch.compile` by replacing the NumPy import by `import torch._numpy as np`.
```
Actual question: is torch.compile(backend="eager") equivalent?

lezcano replied:
Almost!
The flow chart is:
- If it fails with `torch._numpy as np` it means that the error is in the decomposition
- If it doesn't, but fails on `torch.compile(backend="eager")`, the error is in dynamo
- If it doesn't, but fails on `torch.compile`, the error is in inductor

The story in the post does not split the second and third case, as the user doesn't care about where in `torch.compile` the error is.

ev-br commented on blogpost/post.md:
```diff
@@ -0,0 +1,261 @@
+replacement for the PyTorch API, as it is **much slower** and, as a private API,
+**may change without notice**.
+
+## Differences between NumPy and `torch.compile`d NumPy
```
May want to link to https://github.com/Quansight-Labs/numpy_pytorch_interop/issues/5 (or move it somewhere)

lezcano replied:
Yes, that's on my TODO list for next week!

ev-br commented on blogpost/post.md:
````diff
@@ -0,0 +1,261 @@
+times, a bit surprising
+```python
+>>> np.asarray([1], dtype=np.int8) + 127
+array([128], dtype=int8)
````
It is -128 for both new and legacy promotion?

```
In [9]: np._set_promotion_state("weak")

In [10]: np.asarray([1], dtype=np.int8) + 127
Out[10]: array([-128], dtype=int8)

In [11]: np._set_promotion_state("legacy")

In [12]: np.asarray([1], dtype=np.int8) + 127
Out[12]: array([-128], dtype=int8
```

which agrees with

```
In [45]: torch.as_tensor(1, dtype=torch.int8) + 127
Out[45]: tensor(-128, dtype=torch.int8)
```

lezcano replied:
Oops. This should be `+ 126` not to hit an overflow. Good catch!

ev-br commented:
Reads great! 
I've left several comments, all minor, can be safely ignored.

My main confusion is around the section about mixing numpy arrays and torch tensors. Might be great to also stress a recommendation for mixing the two: does torch.compile gracefully handle the mix, what is the recommendation (do not mix, mix freely or something else).

lezcano commented:
Merging just to archive it. The last version of this post is at https://hackmd.io/@82vuWEfETza6QgoTVJYnew/SkGEpt_R2/edit
